src/main/java/sparksoniq/jsoniq/runtime/iterator/operational/TypePromotionIterator.java

```diff
@@ -0,0 +1,173 @@
+import sparksoniq.jsoniq.runtime.metadata.IteratorMetadata;
+import sparksoniq.semantics.DynamicContext;
+import sparksoniq.semantics.types.ItemType;
+import sparksoniq.semantics.types.ItemTypes;
+import sparksoniq.semantics.types.SequenceType;
+
+import java.util.Collections;
+
+public class TypePromotionIterator extends HybridRuntimeIterator {
+
+    private static final long serialVersionUID = 1L;
+    private final String _exceptionMessage;
+    private RuntimeIterator _iterator;
+    private SequenceType _sequenceType;
+
+    private ItemType itemType;
+    private String sequenceTypeName;
+
+    private Item _nextResult;
+    private int _childIndex;
+
+    public TypePromotionIterator(
+            String exceptionMessage,
+            RuntimeIterator iterator,
+            SequenceType sequenceType,
+            IteratorMetadata iteratorMetadata
+    ) {
+        super(Collections.singletonList(iterator), iteratorMetadata);
+        this._exceptionMessage = exceptionMessage;
+        this._iterator = iterator;
+        this._sequenceType = sequenceType;
+        itemType = _sequenceType.getItemType();
+        sequenceTypeName = ItemTypes.getItemTypeName(itemType.getType().toString());
+    }
+
+    @Override
+    public boolean hasNextLocal() {
+        return _hasNext;
+    }
+
+    @Override
+    public void resetLocal(DynamicContext context) {
+        _iterator.reset(_currentDynamicContext);
+        setNextResult();
+    }
+
+    @Override
+    public void closeLocal() {
+        _iterator.close();
+    }
+
+    @Override
+    public void openLocal() {
+        this._childIndex = 0;
+        _iterator.open(_currentDynamicContext);
+        this.setNextResult();
+    }
+
+
+    @Override
+    public Item nextLocal() {
+        if (this._hasNext) {
+            Item _currentResult = _nextResult;
+            setNextResult();
+            return _currentResult;
+        } else
+            throw new IteratorFlowException(RuntimeIterator.FLOW_EXCEPTION_MESSAGE, getMetadata());
+    }
+
+    private void setNextResult() {
+        _nextResult = null;
+        if (_iterator.hasNext()) {
+            _nextResult = _iterator.next();
+            if (_nextResult != null)
+                _childIndex++;
+        } else {
+            _iterator.close();
+            checkEmptySequence(_childIndex);
+        }
+
+        this._hasNext = _nextResult != null;
+        if (!hasNext())
+            return;
+
+        checkItemsSize(_childIndex);
+        if (!_nextResult.isTypeOf(itemType)) {
+            checkTypePromotion();
+        }
+    }
+
+    private void checkEmptySequence(long size) {
+        if (
+            size == 0
+                && (_sequenceType.getArity() == SequenceType.Arity.One
+                    ||
+                    _sequenceType.getArity() == SequenceType.Arity.OneOrMore)
+        ) {
+            throw new UnexpectedTypeException(
+                    "Empty sequence cannot be promoted to type "
```

Comment:
The string in the error message is not quite correct: only atomic values can be promoted ( https://www.w3.org/TR/xquery-30/#id-type-promotion-and-operator-mapping ). If an empty sequence is found and the arity is at least 1, then this is a type error.

How about : "Expecting at least one item, but the value provided is the empty sequence." or along these lines?

Reply:
I am fine with that as well. I just want to point out that Zorba is throwing 

> type error [err:XPTY0004]: empty-sequence() can not be promoted to parameter type xs:integer of function round-half-to-even()

when evaluating `round-half-to-even(1.125, ())` and

> type error [err:XPTY0004]: sequence of more than one item can not be promoted to parameter type xs:integer of function round-half-to-even()

when evaluating `round-half-to-even(1.125, (1,2,3))`.

If you feel that this is not entirely correct, I will proceed with the message you suggested.

Reply:
Thank you for the thorough check with Zorba. It is always a good idea to double-check.

I looked at the specification again and think the error message in Zorba was an abuse of language. I cannot find any mention in the spec of the promotion of sequences.

It would be best to not follow Zorba here, and output error messages that do not use the term "promotion", excepted when referring to a single atomic item and atomic type (which is the case for one message).